usd/src/UsdError.cc

```diff
@@ -0,0 +1,176 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "sdf/usd/UsdError.hh"
+
+#include <iostream>
+
+using namespace sdf::usd;
+
+class sdf::usd::UsdError::Implementation
+{
+  /// \brief The error code value.
+  public: UsdErrorCode code = UsdErrorCode::NONE;
+
+  /// \brief Description of the error.
+  public: std::string message = "";
+
+  /// \brief File path where the error was raised.
+  public: std::optional<std::string> filePath = std::nullopt;
+
+  /// \brief Line number in the file path where the error was raised.
+  public: std::optional<int> lineNumber = std::nullopt;
+
+  /// \brief SDF error
+  public: std::optional<sdf::Error> sdf_error = std::nullopt;
+};
+
+/////////////////////////////////////////////////
+UsdError::UsdError()
+  : dataPtr(ignition::utils::MakeImpl<Implementation>())
+{
+}
+
+/////////////////////////////////////////////////
+UsdError::UsdError(const UsdErrorCode _code, const std::string &_message)
+    : dataPtr(ignition::utils::MakeImpl<Implementation>())
+{
+  this->dataPtr->code = _code;
+  this->dataPtr->message = _message;
+}
+
+/////////////////////////////////////////////////
+UsdError::UsdError(const UsdErrorCode _code, const std::string &_message,
+                   const std::string &_filePath)
+    : dataPtr(ignition::utils::MakeImpl<Implementation>())
+{
+  this->dataPtr->code = _code;
+  this->dataPtr->message = _message;
+  this->dataPtr->filePath = _filePath;
+}
+
+/////////////////////////////////////////////////
+UsdError::UsdError(const UsdErrorCode _code, const std::string &_message,
+                   const std::string &_filePath, int _lineNumber)
+    : dataPtr(ignition::utils::MakeImpl<Implementation>())
+{
+  this->dataPtr->code = _code;
+  this->dataPtr->message = _message;
+  this->dataPtr->filePath = _filePath;
+  this->dataPtr->lineNumber = _lineNumber;
+}
+
+/////////////////////////////////////////////////
+UsdError::UsdError(const sdf::Error &_sdf_error)
+    : dataPtr(ignition::utils::MakeImpl<Implementation>())
+{
+  this->dataPtr->code = UsdErrorCode::SDF_ERROR;
+  this->dataPtr->sdf_error = _sdf_error;
+}
+
+/////////////////////////////////////////////////
+UsdErrorCode UsdError::Code() const
+{
+  return this->dataPtr->code;
+}
+
+/////////////////////////////////////////////////
+std::string UsdError::Message() const
+{
+  return this->dataPtr->message;
+}
+
+/////////////////////////////////////////////////
+std::optional<std::string> UsdError::FilePath() const
```

Comment:
Since the string here is stored in a `std::optional`, it looks like the only option we have is to use [`std::reference_wrapper`](https://en.cppreference.com/w/cpp/utility/functional/reference_wrapper) ([source](https://stackoverflow.com/questions/63773503/can-i-somehow-return-stdoptional-with-const-reference-to-my-data-to-avoid-copy)). This would mean that users have to make an extra `get()` call in order to get the actual const ref string. You can see how this would look in the tests:

<details>
  <summary>std::reference_wrapper diff</summary>

```diff
diff --git a/usd/include/sdf/usd/UsdError.hh b/usd/include/sdf/usd/UsdError.hh
index 68cc1f4e..10646ee7 100644
--- a/usd/include/sdf/usd/UsdError.hh
+++ b/usd/include/sdf/usd/UsdError.hh
@@ -17,6 +17,7 @@
 #ifndef SDF_USD_USDERROR_HH_
 #define SDF_USD_USDERROR_HH_
 
+#include <functional>
 #include <optional>
 #include <string>
 #include <vector>
@@ -108,7 +109,8 @@ namespace sdf
     /// \brief Get the file path associated with this error.
     /// \return Returns the path of the file that this error is related to,
     /// nullopt otherwise.
-    public: std::optional<std::string> FilePath() const;
+    public: std::optional<std::reference_wrapper<const std::string>>
+            FilePath() const;
 
     /// \brief Sets the file path that is associated with this error.
     /// \param[in] _filePath The file path that is related to this error. (e.g.
diff --git a/usd/src/UsdError.cc b/usd/src/UsdError.cc
index 0d31b94f..2d0f76ff 100644
--- a/usd/src/UsdError.cc
+++ b/usd/src/UsdError.cc
@@ -17,6 +17,7 @@
 
 #include "sdf/usd/UsdError.hh"
 
+#include <functional>
 #include <iostream>
 
 using namespace sdf::usd;
@@ -95,7 +96,8 @@ const std::string &UsdError::Message() const
 }
 
 /////////////////////////////////////////////////
-std::optional<std::string> UsdError::FilePath() const
+std::optional<std::reference_wrapper<const std::string>>
+UsdError::FilePath() const
 {
   return this->dataPtr->filePath;
 }
diff --git a/usd/src/UsdError_TEST.cc b/usd/src/UsdError_TEST.cc
index 9c7a8768..caeee345 100644
--- a/usd/src/UsdError_TEST.cc
+++ b/usd/src/UsdError_TEST.cc
@@ -37,7 +37,7 @@ TEST(Error, DefaultConstruction)
   }
   error.SetFilePath("/usd/world");
   ASSERT_TRUE(error.FilePath().has_value());
-  EXPECT_EQ("/usd/world", error.FilePath());
+  EXPECT_EQ("/usd/world", error.FilePath().value().get());
 
   error.SetLineNumber(5);
   ASSERT_TRUE(error.LineNumber().has_value());
@@ -73,7 +73,7 @@ TEST(Error, ValueConstructionWithFile)
   EXPECT_EQ(error.Code(), sdf::usd::UsdErrorCode::INVALID_PRIM_PATH);
   EXPECT_EQ(error.Message(), "Invalid Prim path");
   ASSERT_TRUE(error.FilePath().has_value());
-  EXPECT_EQ(error.FilePath().value(), emptyFilePath);
+  EXPECT_EQ(error.FilePath().value().get(), emptyFilePath);
   EXPECT_FALSE(error.LineNumber().has_value());
   EXPECT_FALSE(error.SdfError().has_value());
 
@@ -95,7 +95,7 @@ TEST(Error, ValueConstructionWithLineNumber)
   EXPECT_EQ(error.Code(), sdf::usd::UsdErrorCode::INVALID_PRIM_PATH);
   EXPECT_EQ(error.Message(), "Invalid Prim path");
   ASSERT_TRUE(error.FilePath().has_value());
-  EXPECT_EQ(error.FilePath().value(), emptyFilePath);
+  EXPECT_EQ(error.FilePath().value().get(), emptyFilePath);
   ASSERT_TRUE(error.LineNumber().has_value());
   EXPECT_EQ(error.LineNumber().value(), lineNumber);
   EXPECT_FALSE(error.SdfError().has_value());
```

</details>

Do you think this is a good approach to take? Or should we leave the method signature as-is, returning a `std::optional<std::string>`?

Reply:
Oh, I meant `const std::optional<std::string> &`. Sorry for not being clear.

Reply:
No problem, yeah we can definitely do that - I made the change in https://github.com/ignitionrobotics/sdformat/pull/836/commits/4be0adaa92bc12505a14a647cc2183b0079ec8c3